File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/BT/SubsetSum.swift
import Foundation

func printSubset(_ flags : inout [Bool], _ arr : inout [Int], _ size : Int) {
	var i : Int = 0
	while (i < size) {
		if (flags[i]) {
			print(String(arr[i]), terminator: " ")
		}
		i += 1
	}
	print()
}

func subsetSum(_ arr : inout [Int], _ n : Int, _ target : Int) {
	var flags : [Bool] = Array(repeating: false, count: n)
	subsetSum( &arr,n, &flags,0,0,target)
}

func subsetSum(_ arr : inout [Int], _ n : Int, _ flags : inout [Bool], _ sum : Int, _ curr : Int, _ target : Int) {
	if (target == sum) { // Solution found.
		printSubset( &flags, &arr,n)
		return
	}

	if (curr >= n || sum > target) { // constraint check and Backtracking.
		return
	}
	
	flags[curr] = true // Current element included.
	subsetSum( &arr,n, &flags,sum + arr[curr],curr + 1,target)
	flags[curr] = false // Current element excluded.
	subsetSum( &arr,n, &flags,sum,curr + 1,target)
}

// Testing code.
var arr : [Int] = [15, 22, 14, 26, 32, 9, 16, 8]
let target : Int = 53
let n : Int = arr.count
subsetSum( &arr,n,target)

/*
15 22 16 
15 14 16 8 
22 14 9 8 
*/

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/BT/SubsetSum.ts
function printSubset(flags: boolean[], arr: number[], size: number): void {
    let result = "";
    for (let i = 0; i < size; i++) {
        if (flags[i]) {
            result += arr[i];
            result += " ";
        }
    }
    console.log(result);
}

function subsetSum(arr: number[], n: number, target: number): void {
    const flags: boolean[] = new Array(n).fill(false);
    subsetSumUtil(arr, n, flags, 0, 0, target);
}

function subsetSumUtil(arr: number[], n: number, flags: boolean[], sum: number, curr: number, target: number): void {
    if (target === sum) {
        printSubset(flags, arr, n); // Solution found.
        return;
    }

    if (curr >= n || sum > target) {
        // Constraint check and backtracking.
        return;
    }

    // Current element included.
    flags[curr] = true;
    subsetSumUtil(arr, n, flags, sum + arr[curr], curr + 1, target);
    flags[curr] = false; // Reset flag for backtracking.
    subsetSumUtil(arr, n, flags, sum, curr + 1, target);
}

// Testing code.
const arr: number[] = [15, 22, 14, 26, 32, 9, 16, 8];
const target: number = 53;
subsetSum(arr, arr.length, target);

/*
15 22 16 
15 14 16 8 
22 14 9 8 
*/
